### PR TITLE
Dispatch events (to get a fancy decoupled output)

### PR DESCRIPTION
**Pull request by jjanvier**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -26,7 +30,7 @@
             "Akeneo\\CouplingDetector\\": "src/Akeneo/CouplingDetector/"
         }
     },
-    "bin": ["bin/coupling-detector"],
+    "bin": ["bin/php-coupling-detector"],
```
Comment (nidup) [inline composer.json]: yep sounds better

File: src/Akeneo/CouplingDetector/Event/Events.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    const POST_NODES_PARSED = 'php_cd.post_nodes_parsed';
+
+    /**
+     * The PRE_RULES_CHECKED event occurs before all have been checked.
+     *
+     * The event listener method receives a Akeneo\CouplingDetector\Event\PreRulesCheckedEvent instance.
+     *
+     * @Event
+     */
+    const PRE_RULES_CHECKED = 'php_cd.pre_rules_checked';
```
Comment (nidup) [inline src/Akeneo/CouplingDetector/Event/Events.php]: for a rule i would maybe say "matched" and not "checked", it imho seems more meaningful

File: composer.json [hunk 3]
```diff
@@ -13,10 +13,14 @@
...
-        "fabpot/php-cs-fixer": "1.11.*"
+        "fabpot/php-cs-fixer": "~1.11",
+        "symfony/event-dispatcher": "~2.3|~3.0",
+        "symfony/finder": "~2.3|~3.0",
+        "symfony/console": "~2.3|~3.0",
+        "symfony/filesystem": "~2.3|~3.0"
     },
     "require-dev": {
-        "phpspec/phpspec": "2.3.*"
+        "phpspec/phpspec": "~2.3"
```
Comment (nidup) [inline composer.json]: we can bump to 2.5, released today :)

File: src/Akeneo/CouplingDetector/Formatter/AbstractFormatter.php [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        $this->outputPostRulesChecked($event);
+    }
+
+    protected abstract function outputPreNodesParsed(PreNodesParsedEvent $event);
+    protected abstract function outputNodeParsed(NodeParsedEvent $event);
+    protected abstract function outputPostNodesParsed(PostNodesParsedEvent $event);
+    protected abstract function outputPreRulesChecked(PreRulesCheckedEvent $event);
+    protected abstract function outputNodeChecked(NodeChecked $event);
+    protected abstract function outputRuleChecked(RuleCheckedEvent $event);
+    protected abstract function outputPostRulesChecked(PostRulesCheckedEvent $event);
```
Comment (nidup) [inline src/Akeneo/CouplingDetector/Formatter/AbstractFormatter.php]: Interesting way to define the contract that need to be implemented by your formatter.

I wonder if this should be the responsibility of another interface, let's say
```
interface FormatterInterface {
    public function outputPreNodesParsed(PreNodesParsedEvent $event);
    public function outputNodeParsed(NodeParsedEvent $event);
    public function outputPostNodesParsed(PostNodesParsedEvent $event);
    public function outputPreRulesChecked(PreRulesCheckedEvent $event);
    public function outputNodeChecked(NodeChecked $event);
    public function outputRuleChecked(RuleCheckedEvent $event);
    public function outputPostRulesChecked(PostRulesCheckedEvent $event);
}

An implementation of this interface should be injected in a FormatterSubscriber.

It looks a bit weird to me to say that a Formatter is a Subscriber in fact :)

**Comment by nidup** [comment]
Very serious OOP here :wink: 
Such great also because it gives me bunch of idea about how we should decouple and implement several PIM behaviors based on a proper event listener system like this one.
Write down some questions about Formatter + Subscriber but it's GTM for me.

File: composer.json (hunk 3 above)
Comment (jjanvier) [inline composer.json]: ~2.3 should take 2.5 no?

Comment (nidup) [inline composer.json]: True, i misread :)

**Pull request merged**